File: src/lib/extractor.ts
import type { EmailSample, ExtractionCandidate, ResultType } from "./types";

const CODE_PATTERN = /(?<![a-z0-9])([a-z0-9]{4,8}|\d{4,8})(?![a-z0-9])/gi;
const LINK_PATTERN = /https?:\/\/[^\s"'<>]+/gi;
const HREF_PATTERN = /href=["'](https?:\/\/[^"']+)["']/gi;
const TRACKING_PARAMS = new Set(["utm_source", "utm_medium", "utm_campaign", "utm_term", "utm_content", "fbclid", "gclid"]);

export function extractCandidates(sample: EmailSample): ExtractionCandidate[] {
  const seen = new Set<string>();
  const candidates: ExtractionCandidate[] = [];

  collectCodes(sample.subject, "subject", candidates, seen, 96);
  collectCodes(sample.bodyText, "body", candidates, seen, 84);
  collectLinks(sample.bodyText, "body", candidates, seen, 88);
  collectHtmlLinks(sample.bodyHtml ?? "", candidates, seen);

  return candidates.sort((left, right) => right.confidence - left.confidence || left.value.localeCompare(right.value));
}

export function maskSecretPreview(value: string, type: ResultType): string {
  if (type === "code") {
    if (value.length <= 4) return "*".repeat(value.length);
    return `${value.slice(0, 2)}****${value.slice(-2)}`;
  }

  try {
    const url = new URL(value);
    return `${url.origin}${url.pathname}?...`;
  } catch {
    return value.length > 18 ? `${value.slice(0, 18)}...` : value;
  }
}

export function sanitizeLink(value: string): string {
  const decoded = decodeHtmlEntities(value).replace(/[).,;]+$/g, "");

  try {
    const url = new URL(decoded);
    for (const key of [...url.searchParams.keys()]) {
      if (TRACKING_PARAMS.has(key.toLowerCase())) {
        url.searchParams.delete(key);
      }
    }
    return url.toString();
  } catch {
    return decoded;
  }
}

function collectCodes(
  value: string,
  source: ExtractionCandidate["source"],
  candidates: ExtractionCandidate[],
  seen: Set<string>,
  baseConfidence: number,
) {
  const searchable = stripLinks(value);
  for (const match of searchable.matchAll(CODE_PATTERN)) {
    const code = match[1];
    if (!isLikelyCode(code)) continue;

    const key = `code:${code.toLowerCase()}`;
    if (seen.has(key)) continue;
    seen.add(key);

    candidates.push({
      type: "code",
      value: code,
      preview: maskSecretPreview(code, "code"),
      source,
      confidence: baseConfidence + codeContextBoost(value, match.index ?? 0),
    });
  }
}

function collectLinks(
  value: string,
  source: ExtractionCandidate["source"],
  candidates: ExtractionCandidate[],
  seen: Set<string>,
  baseConfidence: number,
) {
  for (const match of value.matchAll(LINK_PATTERN)) {
    pushLinkCandidate(match[0], source, candidates, seen, baseConfidence);
  }
}

function collectHtmlLinks(html: string, candidates: ExtractionCandidate[], seen: Set<string>) {
  for (const match of html.matchAll(HREF_PATTERN)) {
    const linkText = extractAnchorText(html, match.index ?? 0);
    const boost = /verify|confirm|approve|login|sign|auth|认证|验证|登录/i.test(linkText) ? 10 : 0;
    pushLinkCandidate(match[1], "html", candidates, seen, 90 + boost);
  }
}

function pushLinkCandidate(
  rawValue: string,
  source: ExtractionCandidate["source"],
  candidates: ExtractionCandidate[],
  seen: Set<string>,
  confidence: number,
) {
  const link = sanitizeLink(rawValue);
  const key = `link:${link}`;
  if (seen.has(key)) return;
  seen.add(key);

  candidates.push({
    type: "link",
    value: link,
    preview: maskSecretPreview(link, "link"),
    source,
    confidence,
  });
}

function isLikelyCode(value: string): boolean {
  if (/^\d{4,8}$/.test(value)) return true;
  if (/^(?=.*\d)(?=.*[a-z])[a-z0-9]{6,8}$/i.test(value)) return true;
  return false;
}

function codeContextBoost(text: string, index: number): number {
  const window = text.slice(Math.max(0, index - 24), index + 32).toLowerCase();
  if (/code|验证码|verification|verify|otp|pin/.test(window)) return 8;
  return 0;
}

function decodeHtmlEntities(value: string): string {
  return value.replace(/&amp;/g, "&").replace(/&quot;/g, '"').replace(/&#39;/g, "'").replace(/&lt;/g, "<").replace(/&gt;/g, ">");
}

function extractAnchorText(html: string, hrefIndex: number): string {
  const close = html.indexOf("</a>", hrefIndex);
  const openEnd = html.indexOf(">", hrefIndex);
  if (close === -1 || openEnd === -1 || close <= openEnd) return "";
  return html.slice(openEnd + 1, close);
}

function stripLinks(value: string): string {
  return value.replace(LINK_PATTERN, " ");
}

File: src-tauri/src/extractor.rs
use crate::types::{EmailSample, ExtractionCandidate, ResultType};
use regex::Regex;
use std::collections::HashSet;

pub fn extract_candidates(sample: &EmailSample) -> Vec<ExtractionCandidate> {
    let mut seen = HashSet::new();
    let mut candidates = Vec::new();

    collect_codes(&sample.subject, "subject", 96, &mut seen, &mut candidates);
    collect_codes(&strip_links(&sample.body_text), "body", 84, &mut seen, &mut candidates);
    collect_links(&sample.body_text, "body", 88, &mut seen, &mut candidates);
    collect_html_links(sample.body_html.as_deref().unwrap_or_default(), &mut seen, &mut candidates);

    candidates.sort_by(|left, right| right.confidence.cmp(&left.confidence).then(left.value.cmp(&right.value)));
    candidates
}

pub fn mask_secret_preview(value: &str, result_type: &ResultType) -> String {
    match result_type {
        ResultType::Code => {
            if value.chars().count() <= 4 {
                "*".repeat(value.chars().count())
            } else {
                format!("{}****{}", &value[..2], &value[value.len() - 2..])
            }
        }
        ResultType::Link => url::Url::parse(value)
            .map(|url| format!("{}{}?...", url.origin().unicode_serialization(), url.path()))
            .unwrap_or_else(|_| {
                if value.len() > 18 {
                    format!("{}...", &value[..18])
                } else {
                    value.to_string()
                }
            }),
    }
}

pub fn sanitize_link(value: &str) -> String {
    let trimmed = decode_entities(value).trim_end_matches([')', '.', ',', ';']).to_string();
    let Ok(mut url) = url::Url::parse(&trimmed) else {
        return trimmed;
    };

    let tracking = ["utm_source", "utm_medium", "utm_campaign", "utm_term", "utm_content", "fbclid", "gclid"];
    let pairs = url
        .query_pairs()
        .filter(|(key, _)| !tracking.contains(&key.to_lowercase().as_str()))
        .map(|(key, value)| (key.to_string(), value.to_string()))
        .collect::<Vec<_>>();
    url.query_pairs_mut().clear().extend_pairs(pairs);
    url.to_string()
}

fn collect_codes(
    text: &str,
    source: &str,
    base_confidence: i32,
    seen: &mut HashSet<String>,
    candidates: &mut Vec<ExtractionCandidate>,
) {
    let regex = Regex::new(r"(?i)(^|[^a-z0-9])([a-z0-9]{4,8}|\d{4,8})([^a-z0-9]|$)").expect("valid code regex");
    for capture in regex.captures_iter(text) {
        let Some(value) = capture.get(2).map(|item| item.as_str()) else {
            continue;
        };
        if !is_likely_code(value) {
            continue;
        }
        let key = format!("code:{}", value.to_lowercase());
        if !seen.insert(key) {
            continue;
        }
        candidates.push(ExtractionCandidate {
            result_type: ResultType::Code,
            value: value.to_string(),
            preview: mask_secret_preview(value, &ResultType::Code),
            source: source.to_string(),
            confidence: base_confidence,
        });
    }
}

fn collect_links(
    text: &str,
    source: &str,
    base_confidence: i32,
    seen: &mut HashSet<String>,
    candidates: &mut Vec<ExtractionCandidate>,
) {
    let regex = Regex::new(r#"https?://[^\s"'<>]+"#).expect("valid link regex");
    for match_item in regex.find_iter(text) {
        push_link(match_item.as_str(), source, base_confidence, seen, candidates);
    }
}

fn collect_html_links(html: &str, seen: &mut HashSet<String>, candidates: &mut Vec<ExtractionCandidate>) {
    let regex = Regex::new(r#"href=["'](https?://[^"']+)["']"#).expect("valid href regex");
    for capture in regex.captures_iter(html) {
        if let Some(value) = capture.get(1).map(|item| item.as_str()) {
            push_link(value, "html", 90, seen, candidates);
        }
    }
}

fn push_link(
    value: &str,
    source: &str,
    confidence: i32,
    seen: &mut HashSet<String>,
    candidates: &mut Vec<ExtractionCandidate>,
) {
    let link = sanitize_link(value);
    let key = format!("link:{link}");
    if !seen.insert(key) {
        return;
    }
    candidates.push(ExtractionCandidate {
        result_type: ResultType::Link,
        preview: mask_secret_preview(&link, &ResultType::Link),
        value: link,
        source: source.to_string(),
        confidence,
    });
}

fn is_likely_code(value: &str) -> bool {
    Regex::new(r"^\d{4,8}$").expect("valid numeric regex").is_match(value)
        || Regex::new(r"(?i)^(?=.*\d)(?=.*[a-z])[a-z0-9]{6,8}$")
            .expect("valid alphanumeric regex")
            .is_match(value)
}

fn strip_links(value: &str) -> String {
    Regex::new(r#"https?://[^\s"'<>]+"#)
        .expect("valid strip regex")
        .replace_all(value, " ")
        .to_string()
}

fn decode_entities(value: &str) -> String {
    value
        .replace("&amp;", "&")
        .replace("&quot;", "\"")
        .replace("&#39;", "'")
        .replace("&lt;", "<")
        .replace("&gt;", ">")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn extracts_code_and_sanitized_link() {
        let sample = EmailSample {
            sender: "AWS".to_string(),
            subject: "Code 839204".to_string(),
            body_text: "Use 839204 at https://example.com/verify?token=abc&utm_source=x".to_string(),
            body_html: None,
        };
        let candidates = extract_candidates(&sample);
        assert!(candidates.iter().any(|item| item.value == "839204"));
        assert!(candidates.iter().any(|item| item.value == "https://example.com/verify?token=abc"));
    }
}
